src/SkyCoords.jl

```diff
@@ -158,29 +158,32 @@ coords2cart(c::AbstractSkyCoords) = coords2cart(lon(c), lat(c))
 rotmat{T1<:GalCoords, T2<:ICRSCoords}(::Type{T1}, ::Type{T2}) = ICRS_TO_GAL
 rotmat{T1<:ICRSCoords, T2<:GalCoords}(::Type{T1}, ::Type{T2}) = GAL_TO_ICRS
 
+getequinox{eq}(x::Type{Type{FK5Coords{eq}}}) = eq
+getequinox{eq,T}(x::Type{Type{FK5Coords{eq,T}}}) = eq
+
 # Define both these so that `convert(FK5Coords{e}, ...)` and
 # `convert(FK5Coords{e,T}, ...)` both work. Similar with other
 # FK5Coords rotmat methods below.
-rotmat{e1, T2<:ICRSCoords}(::Type{FK5Coords{e1}}, ::Type{T2}) =
-    precess_from_j2000(e1) * ICRS_TO_FK5J2000
-rotmat{e1, T1, T2<:ICRSCoords}(::Type{FK5Coords{e1,T1}}, ::Type{T2}) =
-    precess_from_j2000(e1) * ICRS_TO_FK5J2000
-
-rotmat{e1, T2<:GalCoords}(::Type{FK5Coords{e1}}, ::Type{T2}) =
-    precess_from_j2000(e1) * GAL_TO_FK5J2000
-rotmat{e1, T1, T2<:GalCoords}(::Type{FK5Coords{e1,T1}}, ::Type{T2}) =
-    precess_from_j2000(e1) * GAL_TO_FK5J2000
-
-rotmat{T1<:ICRSCoords, e2, T2}(::Type{T1}, ::Type{FK5Coords{e2,T2}}) =
-    FK5J2000_TO_ICRS * precess_from_j2000(e2)'
-
-rotmat{T1<:GalCoords, e2, T2}(::Type{T1}, ::Type{FK5Coords{e2,T2}}) =
-    FK5J2000_TO_GAL * precess_from_j2000(e2)'
-
-rotmat{e1, e2, T2}(::Type{FK5Coords{e1}}, ::Type{FK5Coords{e2,T2}}) =
-    precess_from_j2000(e1) * precess_from_j2000(e2)'
-rotmat{e1, T1, e2, T2}(::Type{FK5Coords{e1,T1}}, ::Type{FK5Coords{e2,T2}}) =
-    precess_from_j2000(e1) * precess_from_j2000(e2)'
+@generated rotmat{E1,T2<:ICRSCoords}(e1::Type{FK5Coords{E1}}, ::Type{T2}) =
+    precess_from_j2000(getequinox(e1)) * ICRS_TO_FK5J2000
```

Comment:
Is `getequinox(e1)` always just `E1` here?

Reply:
No.  This part of the syntax of generated functions may be confusing indeed.

The name of the variable in the body of a generated function is actually its type, not its value.  See the example in the manual: https://docs.julialang.org/en/stable/manual/metaprogramming/#Generated-functions-1.  That's why it can be computed at compile-time rather than at run-time, making generated functions really efficient.  The body of the function is executed only few times during a session (not guaranteed to be run exactly once, but the number is *very* low) and the result is cached.

Look also at `getequinox`, its argument is the type of the type `FK5Coords`, not the type of `FK5Coords` ;-)

In these functions there is no returned expression, which is fine.

Reply:
See for example how many times `getequinox` is called: https://coveralls.io/builds/13759378/source?filename=src%2FSkyCoords.jl#L161

Reply:
Yeah, I get that, but it still seems like `getequinox(e1)` is always `E1`, at least in this method. For example, if `e1` is `Type{FK5Coords{1975}}` within the body of the generated method, then `getequinox(e1)` will be `1975`, which is also the value of `E1` in the body of the generated function. Is that not right?

Reply:
I never realized that type parameters can be used in the body of generated functions 😄  I don't know why, I thought that they weren't accessible.

Reply:
Update the PR, now it simply adds `@generated` decorators before the `rotmat` methods.